EBD-530: Apollo3 support

ianstadler opened the pull request:
Added support for the Apollo3 Blue Plus MCU. This includes the necessary changes to the GPIO, pinctrl, and timer drivers as well as device tree for the Apollo3 Blue Plus development board. This brings the support for this MCU to be at the same level as the support for the Apollo4 Blue Plus in the Zephyr repo.

ianstadler committed:
boards: arm: apollo3p_blue: Added new board

Initial support for the apollo3p blue hardware

Signed-off-by: Ian Stadler <istadler@semios.com>

ianstadler committed:
soc: ambiq: apollo3x: Initial support for Apollo3 Blue Plus

Added using the Ambiq apollo3 HAL along with some basic changes
for the different hardware in the GPIO, STIMER, and PINCTRL
drivers

Signed-off-by: Ian Stadler <istadler@semios.com>

ianstadler committed:
drivers: gpio: ambiq: Added Ambiq Apollo3 support

The apollo3x MCUs only have a single shared interrupt across
all GPIO pins. Changed the driver to handle this difference

Signed-off-by: Ian Stadler <istadler@semios.com>

ianstadler committed:
soc: arm: ambiq: apollo3x: Fixed UART cfg

Updated the UART power configuration to use the actual
apollo3 pwrctrl registers

Signed-off-by: Ian Stadler <istadler@semios.com>

ianstadler committed:
boards: arm: apollo3p_blue_evb: Fixed timer configuration

Updated stimer interrupt from 22 -> 23

Signed-off-by: Ian Stadler <istadler@semios.com>

ianstadler committed:
drivers: gpio: ambiq: Fixed GPIO interrupts

Fixed the configuration of GPIO interrupts so that they are properly
enabled/disabled in the interrupt configuration function rather than
init. Additionally, fixed the get configuration and get direction
API implementations.

Signed-off-by: Ian Stadler <istadler@semios.com>

ianstadler committed:
soc: ambiq: apollo3: Fixed compliance

Fixed several compliance errors

Signed-off-by: Ian Stadler <istadler@semios.com>

ohou-semios commented on dts/arm/ambiq/ambiq_apollo3p_blue.dtsi:
```diff
@@ -0,0 +1,227 @@
+			ambiq,pwrcfg = <&pwrcfg 0x8 0x100>;
+		};
+
+		iom0: iom@40050000 {
```
Just double checking the addresses for the IOM and maybe the stimer.

I may be looking at the wrong datasheet/header file but in `AmbiqSuite_R3.0.0/CMSIS/AmbiqMicro/Include/apollo3p.h` I see no mention an STIMER and also the IOM addresses appear to be different.

```
#define CTIMER_BASE                 0x40008000UL
#define IOM0_BASE                   0x50004000UL
#define IOM1_BASE                   0x50005000UL
#define IOM2_BASE                   0x50006000UL
#define IOM3_BASE                   0x50007000UL
#define IOM4_BASE                   0x50008000UL
#define IOM5_BASE                   0x50009000UL
```

ianstadler replied:
From how they use it in their HAL and examples, the stimer is the same as the ctimer on Apollo3. On Apollo4 they have separate hardware for it.

Good catch on the i2c registers. I was probably looking at the apollo4 data sheet when making the device tree.

ianstadler committed:
dts: arm: ambiq: Fixed IOM MMIO addresses

Was still using the apollo4p addresses for the IOMx
hardware. Fixed this to now properly use the apollo3p
addresses

Signed-off-by: Ian Stadler <istadler@semios.com>
